scripts/build_test.py

```diff
@@ -24,7 +33,79 @@
+        self.assertEqual(calledProcessException.returncode, 1)
+
+    def test_join_files(self, file_count=10):
+        """Determine third_party.js contains the content of the first 10 JS
+        files in /third_party/static.
+        """
+        third_party_js = os.path.join(
+            build.THIRD_PARTY_GENERATED_DEV_DIR, 'js', 'third_party.js')
+        # Get all filepaths from manifest.json.
+        dependency_filepaths = build.get_dependencies_filepaths()
+        # Join all JS files in /third_party/static.
+        # pylint: disable=protected-access
+        build._join_files(dependency_filepaths['js'], third_party_js)
+        # pylint: enable=protected-access
+        with open(third_party_js, 'r') as joined_js_file:
+            # Open third_party.js.
+            joined_js_file_content = joined_js_file.read()
+            counter = 0
+            for js_filepaths in dependency_filepaths['js']:
+                # Open first 10 JS files.
+                if counter == file_count:
+                    break
+                with open(js_filepaths, 'r') as js_file:
+                    # Assert that each line is copied over to third_party.js.
+                    for line in js_file:
+                        self.assertIn(line, joined_js_file_content)
+                counter += 1
+        os.remove(third_party_js)
+
+    def test_minify_and_create_sourcemap(self):
+        """Tests _minify_and_create_sourcemap with an invalid filepath."""
+        # pylint: disable=protected-access
+        with self.assertRaises(subprocess.CalledProcessError) as calledProcess:
+            build._minify_and_create_sourcemap(
+                'path/to/input.js', 'path/to/output.js')
+        # pylint: enable=protected-access
+        calledProcessException = calledProcess.exception
+        # returncode is the exit status of the child process.
+        self.assertEqual(calledProcessException.returncode, 1)
+
+    def test_copy_fonts(self):
+        """Test _copy_fonts to ensure that all fonts are copied over."""
+        # Get all filepaths from manifest.json.
+        dependency_filepaths = build.get_dependencies_filepaths()
+        fonts_dir = os.path.join(
+            build.THIRD_PARTY_GENERATED_STAGING_DIR, 'fonts', '')
+        # pylint: disable=protected-access
+        build._copy_fonts(dependency_filepaths['fonts'], fonts_dir)
+        # pylint: enable=protected-access
+        for fonts_file in dependency_filepaths['fonts']:
+            # Convert wildcard syntax to valid paths.
+            font_paths = glob.glob(fonts_file)
+            for font_path in font_paths:
+                font_name = os.path.basename(font_path)
+                # Assert each font file is copied from /third_party/static
+                # to /backend_prod_files.
+                self.assertTrue(
+                    os.path.isfile(os.path.join(fonts_dir, font_name)))
+        # Clean up /font dir.
+        shutil.rmtree(build.THIRD_PARTY_GENERATED_STAGING_DIR)
```

Comment:
I am a little worried about actually doing file operations in tests. I don't think we should do that. If the tests are incorrect or things change in the core code then it means that the tests will have side-effects.

I would swap out shutil with a mock that ensures that it's called and that the right info is passed in. Similar comments apply in other places.

Reply:
So for copying fonts, how about I setup a specific sandbox folder, test copying fonts in there and then delete that sandbox folder as teardown, how does that sound?

Reply:
All I am looking for is that all fonts in `manifest.json` is accounted for in that test folder. I really think we need to have physical copying done here.

Reply:
After some closer inspection, I will be refactoring _copy_fonts() to just queue up a bunch of copy_tasks and execute those tasks at a later time. Testing will get easier with just trying to calculate how many tasks are supposed to be queued.

Reply:
All done!